File: backends/spring-java/src/main/java/dev/stackverse/backend/message/MessageLocalizer.java
package dev.stackverse.backend.message;

import jakarta.servlet.http.HttpServletRequest;
import org.springframework.http.HttpHeaders;
import org.springframework.stereotype.Component;

@Component
public class MessageLocalizer {
    private final MessageRepository messageRepository;
    private final LanguageResolver languageResolver;

    public MessageLocalizer(MessageRepository messageRepository, LanguageResolver languageResolver) {
        this.messageRepository = messageRepository;
        this.languageResolver = languageResolver;
    }

    public String localize(String key, HttpServletRequest request) {
        String language = languageResolver.resolve(request.getParameter("lang"), request.getHeader(HttpHeaders.ACCEPT_LANGUAGE));
        Message localized = messageRepository.findByKeyAndLanguage(key, language);
        if (localized != null) {
            return localized.getText();
        }
        Message english = messageRepository.findByKeyAndLanguage(key, LanguageResolver.DEFAULT_LANGUAGE);
        return english == null ? key : english.getText();
    }
}

File: backends/spring-kotlin/src/main/kotlin/dev/stackverse/backend/message/MessageLocalizer.kt
package dev.stackverse.backend.message

import jakarta.servlet.http.HttpServletRequest
import org.springframework.http.HttpHeaders
import org.springframework.stereotype.Component

/**
 * Resolves a message key to localized text for the current request (SPEC rule 11):
 * language per rule 8, text from the messages table, `en` fallback, and finally
 * the key itself if no text exists at all.
 */
@Component
class MessageLocalizer(
    private val messageRepository: MessageRepository,
    private val languageResolver: LanguageResolver,
) {

    fun localize(key: String, request: HttpServletRequest): String {
        val language = languageResolver.resolve(
            request.getParameter("lang"),
            request.getHeader(HttpHeaders.ACCEPT_LANGUAGE),
        )
        return messageRepository.findByKeyAndLanguage(key, language)?.text
            ?: messageRepository.findByKeyAndLanguage(key, DEFAULT_LANGUAGE)?.text
            ?: key
    }
}
